src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -175,6 +181,301 @@ static inline ASR::asr_t* create_UnaryFunction(Allocator& al, const Location& lo
+                                int_type, nullptr)),
+                            char_type, nullptr)),
+                            ASR::cmpopType::NotEq,
+                            EXPR(ASR::make_StringItem_t(al, loc, args[0],
+                                EXPR(ASR::make_IntegerBinOp_t(al, loc, i,
+                                    ASR::binopType::Add,
+                                    EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                                    int_type, nullptr)),
+                                char_type, nullptr)),
+                            logical_type, nullptr)),
+                        logical_type, nullptr));
+                    {
+                        Vec<ASR::stmt_t *> if_body_3; if_body_3.reserve(al, 1);
+                        Vec<ASR::stmt_t *> else_body_3; else_body_3.reserve(al, 1);
+                        ASR::expr_t *a_test = EXPR(ASR::make_IntegerCompare_t(al,
+                            loc, j, ASR::cmpopType::NotEq,
+                            EXPR(ASR::make_IntegerConstant_t(al, loc, 0, int_type)),
+                            logical_type, nullptr));
+                        if_body_3.push_back(al, STMT(ASR::make_Assignment_t(al,
+                            loc, j, EXPR(ASR::make_ListItem_t(al, loc, lps,
+                                EXPR(ASR::make_IntegerBinOp_t(al, loc, j,
+                                    ASR::binopType::Sub,
+                                    EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                                    int_type, nullptr)),
+                                int_type, nullptr)), nullptr)));
+                        else_body_3.push_back(al, STMT(ASR::make_Assignment_t(al,
+                            loc, i, EXPR(ASR::make_IntegerBinOp_t(al, loc, i,
+                                ASR::binopType::Add,
+                                EXPR(ASR::make_IntegerConstant_t(al, loc, 1, int_type)),
+                                int_type, nullptr)), nullptr)));
+                        if_body_2.push_back(al, STMT(ASR::make_If_t(al, loc, a_test,
+                            if_body_3.p, if_body_3.n, else_body_3.p, else_body_3.n)));
+                    }
+                    else_body_1.push_back(al, STMT(ASR::make_If_t(al, loc, a_test,
+                        if_body_2.p, if_body_2.n, nullptr, 0)));
+                }
+                loop_body.push_back(al, STMT(ASR::make_If_t(al, loc, a_test,
+                    if_body_1.p, if_body_1.n, else_body_1.p, else_body_1.n)));
+            }
+
+            body.push_back(al, STMT(ASR::make_WhileLoop_t(al, loc, nullptr, a_test,
+                loop_body.p, loop_body.n)));
+        }
+        body.push_back(al, STMT(ASR::make_Return_t(al, loc)));
+
+    }
+    ASR::symbol_t *fn_sym = make_Function_t(fn_name, fn_symtab, dep, args,
+        body, return_var, Source, Implementation, nullptr);
+    global_scope->add_symbol(fn_name, fn_sym);
+    return fn_sym;
+}
+
```

Comment:
Similarly here, it takes about 282 lines of C++ to do what Python can do in 44 lines:

https://github.com/lcompilers/lpython/blob/aa1f1952676d66608b75b18c1c89c66978c87d88/src/runtime/lpython_builtin.py#L713-L757